resources/Microsoft.Windows.Setting.WindowsUpdate/Microsoft.Windows.Setting.WindowsUpdate.psm1

```diff
@@ -0,0 +1,377 @@
+
+.PARAMETER AllowAutoWindowsUpdateDownloadOverMeteredNetwork
+    Indicates whether automatic Windows Update downloads are allowed over metered networks. This is the setting in Windows Update settings -> Advanced options -> Download updates over metered connections.
+
+.PARAMETER RestartNotificationsAllowed
+    Indicates whether restart notifications are allowed. This is the setting in Windows Update settings -> Advanced options -> Notify me when a restart is required to finish updating.
+
+.PARAMETER SmartActiveHoursState
+    Indicates whether smart active hours are enabled.
+
+.PARAMETER UserChoiceActiveHoursEnd
+    The end time for user-chosen active hours.
+
+.PARAMETER UserChoiceActiveHoursStart
+    The start time for user-chosen active hours.
+
+.PARAMETER DownloadMode
+    The download mode for updates. Valid values are 0, 1, and 3. This is the setting in Windows Update settings -> Advanced options -> Delivery Optimization -> Allow downloads from other PCs.
+
+.PARAMETER DownloadRateBackgroundBps
+    The background download rate in bits per second.
+
+.PARAMETER DownloadRateForegroundBps
+    The foreground download rate in bits per second.
+
+.PARAMETER DownloadRateBackgroundPct
+    The background download rate as a percentage.
+
+.PARAMETER DownloadRateForegroundPct
+    The foreground download rate as a percentage.
+
+.PARAMETER UploadLimitGBMonth
+    The upload limit in gigabytes per month.
+
+.PARAMETER UpRatePctBandwidth
+    The upload rate as a percentage of bandwidth.
+
+.EXAMPLE
+    PS C:\> Invoke-DscResource -Name WindowsUpdate -Method Get -ModuleName Microsoft.Windows.Setting.WindowsUpdate -Property @{}
+
+    This command gets the current Windows Update settings.
+#>
+[DSCResource()]
+class WindowsUpdate
+{
+    # Key required. Do not set.
+    [DscProperty(Key)] 
+    [string] $SID
+    
+    [DscProperty()]
+    [nullable[bool]] $IsContinuousInnovationOptedIn
+
+    [DscProperty()]
+    [nullable[bool]] $AllowMUUpdateService
+
+    [DscProperty()]
+    [nullable[bool]] $IsExpedited
+
+    [DscProperty()]
+    [nullable[bool]] $AllowAutoWindowsUpdateDownloadOverMeteredNetwork
+
+    [DscProperty()]
+    [nullable[bool]] $RestartNotificationsAllowed
+
+    [DscProperty()]
+    [nullable[bool]] $SmartActiveHoursState
+
+    [DscProperty()]
+    [nullable[int]] $UserChoiceActiveHoursEnd
+
+    [DscProperty()]
+    [nullable[int]] $UserChoiceActiveHoursStart
+
+    [DscProperty()]
+    [ValidateSet(0, 1, 3)]
+    [nullable[int]] $DownloadMode
+
+    [DscProperty()]
+    [nullable[int]] $DownloadRateBackgroundBps
+
+    [DscProperty()]
+    [nullable[int]] $DownloadRateForegroundBps
+
+    [DscProperty()]
+    [nullable[int]] $DownloadRateBackgroundPct
+
+    [DscProperty()]
+    [nullable[int]] $DownloadRateForegroundPct
+
+    [DscProperty()]
+    [nullable[int]] $UploadLimitGBMonth
+    
+    [DscProperty()]
+    [nullable[int]] $UpRatePctBandwidth
```

Comment:
See above

Reply:
Mhm... I didn't think about that. The only thing why I did such implementation, was of issue #74. Despite not implementing it, if `[Reasons]` is the way forward for these DSC resources, we should provide _reasons_ back. 

This allows for Machine Configuration to capture the actual _reason_ why resources couldn't be applied in the desired state we want, instead of throwing errors. The downside of implementing `[Reasons]` in the DSC resources, is more lines of code. Whether that's a good or bad thing, I'll leave that in the air. 

On that note, I guess the road forward is still far away. I wanted to throw it in there if you didn't know about it :)

Thanks for all the reviews lately Kaleb.

Reply:
Theres no "reason" that you couldn’t do both?  If I'm reading the walkthrough you linked correctly, they have the `ValidateRange` on the `UpdateFrequency` parameter and then later implement `[Reasons]`. I think they are both compatible since the `ValidateRange` only validates what the user input as a parameter, while `[Reasons]` explains why the get method has the results it has

Reply:
You are right, I didn't see that range. I think they are indeed compatible, it is just how we catch it. The difficult part is in the new authoring experience for V3. IntelliSense features are not available (yet). I'll adjust it and remove the functions.